Encode functions

nplrkn opened the pull request:
Hello again, here's a small initial attempt at helping with the encode effort.

- The use of tuples for certain encoded value types was intended to be tidily symmetrical with the result type returned by the decode() equivalent.  

-  I put encode_length_determinent() and the encode charstring fns in mod.rs rather than imitating the current structure of the decode equivalent.  Seems to me that these are all at the same level.  (i.e. they're all called by the codecs_derive derived code).

nplrkn commented:
Mulling over encoding of open types... I think something like this will be needed.

```
impl AperCodecData {
    pub fn length_in_bytes(&self) { ... }
    pub fn append_aligned(&mut self, other: &mut Self) -> Result<(), AperCodecError> {
        todo!()
    }
}
```

gabhijit commented:
> * The use of tuples for certain encoded value types was intended to be tidily symmetrical with the result type returned by the decode() equivalent.

I guess the second `bool` is required if the passed value is not from the `extension_root` and this is more than simply being `is_extensible`. Is that right?

gabhijit commented:
Do you think this is all you want to merge in this pull request? then we can rebase and merge the request and I will pick up from there.

nplrkn commented:
> > * The use of tuples for certain encoded value types was intended to be tidily symmetrical with the result type returned by the decode() equivalent.
> 
> I guess the second `bool` is required if the passed value is not from the `extension_root` and this is more than simply being `is_extensible`. Is that right?

Yes exactly.  Though I'm now going off my original idea of wrapping them in a tuple.

nplrkn commented:
Since posting this PR I have been working on my own project (which uses the codecs module / asn1-codecs crate) to integrate with the encode API.  I just yesterday got to the point where I was starting to look at replacing these stub implementations.   

If it works for you I could extend this PR over the next few days to have a minimal initial implementation.

I'm fully expecting you would want to then refactor and make it how you want it to be.  In parallel I would continue testing using my project and submit fixes for any bugs I find.  What do you think?

gabhijit commented:
> If it works for you I could extend this PR over the next few days to have a minimal initial implementation.

Sure that works.! In fact it will be great if there is some working implementation of encoder. 

> I'm fully expecting you would want to then refactor and make it how you want it to be. In parallel I would continue testing using my project and submit fixes for any bugs I find. What do you think?

Well the whole idea of doing this was it will be useful to others in addition to me, so all the help is always welcome. 

Thanks!

nplrkn commented:
Right, here is my first pass at the encoding functions.  Cargo test is not working at the top level, so that still needs to be fixed.

A few points of interest:

- When encoding, the AperCodecData offset field duplicates bits.len().  So I thought we should make offset a decode specific field and I renamed it as such.
- I noticed a decoding bug that I fixed in passing, related to extracting an unaligned octet string into a Vec<u8>.
- As I hinted earlier, I got fed up with the tuples and got rid of them.
- This doesn't make any attempt yet to cope with extended values.

gabhijit commented:
> Right, here is my first pass at the encoding functions.
> 

Thanks indeed! I was having too much of inertia to get started on this.

> A few points of interest:
> 
>     * When encoding, the `AperCodecData` `offset` field duplicates `bits.len()`.  So I thought we should make `offset` a decode specific field and I renamed it as such.

this is fine. 

> 
>     * I noticed a decoding bug that I fixed in passing, related to extracting an unaligned octet string into a `Vec<u8>`.
> 

Is it possible to add a test case to it? 

>     * As I hinted earlier, I got fed up with the tuples and got rid of them.
> 

Actually to be honest I liked the tuple idea. if you look at current implementation - you are separately sending `extended` flag, which has to be sent (even if not supported immediately.) However, with tuple, it makes more semantically consistent - that is whether the value to be encoded is extended or not passed as a tuple.  But we can revisit that afterwards, for now it is fine.

>     * This doesn't make any attempt yet to cope with extended values.

Do you think it is good enough to be merged or you need more work? 

What I generally prefer with such big pull requests is - I keep the # of commits to 3-4 commits that make sense, rather than adding too many commits to the repository. 

So if that's possible to do it will be great to `rebase` this into a fewer related commits, so better when we merge, this.

Also, one test case I always wanted to add was - Take a bitstream - decode it and encode that structure again, whatever results should be identical to the one originally started with. That's a good test of whether encoder/decoders work.

gabhijit commented:
Hi @nplrkn 

I have marked this PR as a draft one. You can unmark it from the draft one, when you think this is ready to be merged.

nplrkn commented:
Rebased to a single commit and added that test you requested for the unaligned get_bytes().  
Agree that encode/decode tests are a good idea but I didn't add any of those yet.
I think this is reasonably solid and ok to go in.  Expect I'll keep testing it over the next few days.

gabhijit commented:
I believe, let's merge this and then we'll fix any issues that might come up later.

gabhijit commented on codecs_derive/src/aper/choice.rs:
```diff
@@ -84,7 +84,7 @@ fn generate_choice_variant_decode_tokens_using_attrs(
                             let variant_encode_token = quote! {
                                 Self::#variant_ident(ref v) => {
-                                    asn1_codecs::aper::encode::encode_choice_idx(data, #lb, #ub, #ext, #key)?;
+                                    asn1_codecs::aper::encode::encode_choice_idx(data, #lb, #ub, #ext, #key, false)?;
```
I believe this hard-coded `false` is not a good idea here. We are actually taking the `cp.extended.as_ref` above (line #77), we should simply be using that value?

nplrkn replied:
Good point - fixed.

gabhijit commented on codecs/src/aper/mod.rs:
```diff
@@ -221,6 +211,56 @@ impl AperCodecData {
+    }
+
+    /// Add bits to the encoding buffer.
+    fn append_bits(&mut self, bits: &BitSlice<Msb0, u8>) -> Result<(), AperCodecError> {
```
I don't see a reason why this would fail, so may be there is no need to return a `Result`?

nplrkn replied:
Removed the result, and did the same to append_aligned() and encode_bool().

gabhijit reviewed (commented):
I have a couple of small suggestions, can you please take a look at them? Then we can merge this.
